src/CSV.jl

```diff
@@ -30,19 +30,20 @@ include("rows.jl")
 include("write.jl")
 
 """
-`CSV.read(source; copycols::Bool=false, kwargs...)` => `DataFrame`
+`CSV.read(source, sink::T; kwargs...)` => T
 
-Parses a delimited file into a `DataFrame`. `copycols` determines whether a copy of columns should be made when creating the DataFrame; by default, no copy is made.
+Read and parses a delimited file, materializing directly using the `sink` function.
 
-`CSV.read` supports the same keyword arguments as [`CSV.File`](@ref).
+`CSV.read` supports all the same keyword arguments as [`CSV.File`](@ref).
 """
-function read(source; copycols::Bool=false, kwargs...)
-    @warn "`CSV.read(input; kw...)` is deprecated in favor of `DataFrame!(CSV.File(input; kw...))`"
-    DataFrame(CSV.File(source; kwargs...), copycols=copycols)
+function read(source, sink=nothing; copycols::Bool=false, kwargs...)
+    if sink === nothing
+        @warn "`CSV.read(input; kw...)` is deprecated in favor of `CSV.read(input, DataFrame; kw...)"
+        sink = DataFrame
```

Comment:
I understand that with the current design when creating `DataFrame` we do not need to copy columns - right?
```suggestion
        @warn "`CSV.read(input; kw...)` is deprecated in favor of `CSV.read(input, DataFrame!; kw...)"
        sink = DataFrame!
```

Reply:
I thought the point of `CopiedColumns` was to signal to the sink that copying is not necessary (that is, nobody "owns" the columns), so `DataFrame(c::CopiedColumns)` can be defined to be non-copying. If I understand the logic of `fromcolumns`, chances are that should already be the case.

Reply:
@piever is correct; DataFrames.jl already specializes `fromcolumns` for `CopiedColumns` to ensure no additional copies are made; so in this case, it doesn't matter if the user passes `DataFrame` or `DataFrame!`.

Reply:
Ah - right. This is nice indeed.